### PR TITLE
GitHub improvements

### LINKED ISSUE #638: Projects URLs should be able to view any project
Right now one can refer to a view of an already-synced project using an URL like http://app.flowhub.io/#project/rebirth/rebirth_graphs_ExtractColors - however if the project is not synced it will fail. Instead we should be able to on-demand load (a part of) a project.

This can be used for having a "Open in Flowhub" URL in documentation, sending to a friend etc.

It is desirable to be able to specify at least:

* Project/repo
* Branch/commit/tag
* Graph/component/fbpspec/filename

If feasible, we could use the same URL format as on Github. This would make it possible for people to just replace "github.com" with "flowhub.io", to open anything one looks at in Github easily.

As this will be an a feature which external consumers will refer to (in their README.md files, documentation, in emails/chat history etc), it is important that it stays stable. Hence we must have automated tests that exercise the different cases.

### PR DESCRIPTION
**Pull request by bergie**
* [x] Make file operation selector reflect currently selected option
* [x] Fix Polymer notify chain on ops
* [x] Ensure GH sync picks up graph edits as local modifications
* [x] Make pulling project open it
* [x] Throttle GitHub operations so some XHRs don't end up canceled
* [x] If parsing a pulled graph fails, give a friendlier error message
* [x] Add loading state when pulling a project
* [x] Allow pulling projects by URL
* [x] Switch repo checkout UI to use URLs
* [x] Tell anonymous users to log in if their IP is low on API limits

Will example a "Open in Flowhub" mechanism where users can provide a URL to open a given GitHub repo directly in Flowhub with:

* Repo master: https://app.flowhub.io/#github/c-base/ingress-table
* Particular file: https://app.flowhub.io/#github/c-base/ingress-table/blob/master/graphs/FetchDat.json

We also store branch information, but right now that isn't used by the GithubSynchronization graph.

Closes #680
Fixes #638

### CODE REVIEW

**Comment by bergie** [comment]
Now we start with the correct selected operation:
screenshot 2017-11-15 at 16 47 47

**Comment by bergie** [comment]
Now we tell which graph failed to parse, if any:

screenshot 2017-11-15 at 19 17 02

**Comment by bergie** [comment]
So if you provide a `#github/...` URL for repo you don't have yet, you first get a loading banner:

screenshot 2017-11-15 at 20 23 47

...and once the project has been pulled locally, the requested graph opens:

screenshot 2017-11-15 at 20 39 59

**Comment by bergie** [comment]
For anonymous users, GitHub only gives 60 API calls per hour per IP. If user is rate limited, we show this:

screenshot 2017-11-15 at 21 59 35

**Comment by jonnor** [comment]
Is there a particular reason we use the hash for the repo/file path? `#github...`

**Comment by bergie** [comment]
@jonnor we use hash for routing everywhere in the app, since we don't have a server that could serve arbitrary paths

File: components/BlobToEntry.coffee [hunk 1]
```diff
@@ -23,100 +24,132 @@ handleGraph = (sha, content, entry, project, out, callback) ->
...
     project: project.id
     name: entry.remote.name
     code: content
     language: entry.remote.language
     sha: sha
     changed: false
-  project.components.push newEntry
-  out.send newEntry
-  do callback
+  collections.addToList project.components, newEntry
```
Comment (jonnor) [inline components/BlobToEntry.coffee]: What is this `colletions.addToList` thing? If it is just wrapping `list.push()` that serves no purpose and just obscures the code?

Comment (bergie) [reply components/BlobToEntry.coffee]: @jonnor it either pushes, or replaces existing

File: components/PreparePulls.coffee [hunk 2]
```diff
@@ -1,4 +1,24 @@
 noflo = require 'noflo'
+uuid = require 'uuid'
+
+sendPulls = (pulls, repo, scope, output, callback) ->
+  unless pulls.length
+    do callback
+    return
+  entry = pulls.shift()
```
Comment (jonnor) [inline components/PreparePulls.coffee]: Still has mixed app logic and throttling logic

Comment (bergie) [reply components/PreparePulls.coffee]: Yep, let me split to a separate component

Comment (jonnor) [reply components/PreparePulls.coffee]: @bergie just use bluebird.mapSeries or bluebird.map with concurrency param?

Comment (bergie) [reply components/PreparePulls.coffee]: That would involve adding a new dependency

Comment (bergie) [reply components/PreparePulls.coffee]: On another thought, let me keep it as-is for now. We should figure out a proper way to do throttling with Process API on NoFlo level

NoFlo knows the "load" of the component, but unlike legacy AsyncComponent, we're not exposing this to the network in any way

Comment (jonnor) [reply components/PreparePulls.coffee]: If it is not important to show partial results, I don't think there are any benefits from doing throttling on NoFlo level

Comment (bergie) [reply components/PreparePulls.coffee]: Some notes: https://github.com/noflo/noflo/issues/581

**Pull request merged**